src/Logging.js

```diff
@@ -0,0 +1,380 @@
+/**
+ * Client Side Logging module.
+ * @module Logging
+ */
+
+// Note that we have an excessive amount of try/catch blocks in this code.
+// That is intentional. Logging should never break the features which are using it
+
+var TRACE_LEVEL_SILENT = 0;
+var TRACE_LEVEL_ALWAYS = 1;
+var TRACE_LEVEL_ERROR = 2;
+var TRACE_LEVEL_WARN = 3;
+var TRACE_LEVEL_INFO = 4;
+var TRACE_LEVEL_LOG = 5;
+var TRACE_LEVEL_DEBUG = 6;
+var TRACE_LEVEL_VERBOSE = 6;
+
+var LOCAL_STORAGE_KEY_NAME = 'AppNexus_Page_Debug_Log_Level';
+
+// not an actual debug level, but used as a shortcut so we always
+// know what the max debug level is
+var TRACE_LEVEL_ALL = TRACE_LEVEL_DEBUG;
+var TRACE_LEVEL_DEFAULT = TRACE_LEVEL_SILENT;
+
+// the current debug level to use
+var _curDebugLevel = TRACE_LEVEL_DEFAULT;
+
+// the debug level that was set via querystring
+var _debugLevelQueryString = TRACE_LEVEL_DEFAULT;
+
+// the debug level that was set via localStorage
+var _debugLevelLocalStorage = TRACE_LEVEL_DEFAULT;
+
+// the debug level that was set via call to setDebugLevel
+var _debugLevelFunctionSet = TRACE_LEVEL_DEFAULT;
+
+function getCurrentTimeString() {
+    var dateToReturn = '';
+    try {
+        var curDate = new Date();
+        dateToReturn = curDate.getHours() + ':' + curDate.getMinutes() + ':' + curDate.getSeconds() + '.' + curDate.getMilliseconds();
+    } catch (e) {}
+    return dateToReturn;
+}
+
+function getTraceMethodName(messageLogLevel) {
```

Comment:
is it necessary to have this many debug levels? seems a bit excessive.

Reply:
We use this many debug levels in our other products. These levels correspond to the log levels supported in browser debug console. It helps user to filter log levels in debug console.